### PR TITLE
Avoid using dynamic templates for semantic text fields

### PR DESCRIPTION
**Pull request by carlosdelest**
Avoids `semantic_text` fields to be used in dynamic templates. 

We need the `semantic_text` fields to be already in the mapping so we can perform inference, so we're disallowing it based on the work done in #107491 .

I've refactored some of the existing tests so we can ensure that the field mapper does comply with the requirements, and created some utilities for dealing with inference services.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/index/mapper/NonDynamicFieldMapperTestCase.java
Comment (carlosdelest) [inline server/src/test/java/org/elasticsearch/index/mapper/NonDynamicFieldMapperTestCase.java:1]: Refactored this class to make it a base test case

File: x-pack/plugin/inference/src/internalClusterTest/java/org/elasticsearch/xpack/inference/action/filter/ShardBulkInferenceActionFilterIT.java [hunk 1]
```diff
@@ -143,66 +137,4 @@ public void testBulkOperations() throws Exception {
         }
     }
 
-    private void storeSparseModel() throws Exception {
```
Comment (carlosdelest) [inline x-pack/plugin/inference/src/internalClusterTest/java/org/elasticsearch/xpack/inference/action/filter/ShardBulkInferenceActionFilterIT.java:146]: Refactored out some utility classes to be reused

**Review by jimczi — APPROVED** [review]
LGTM

File: server/src/test/java/org/elasticsearch/index/mapper/NonDynamicFieldMapperTestCase.java [hunk 2]
```diff
@@ -11,114 +11,110 @@
...
-public class DynamicFieldMapperTests extends ESSingleNodeTestCase {
-    @Override
-    protected Collection<Class<? extends Plugin>> getPlugins() {
-        return List.of(NonDynamicFieldPlugin.class);
-    }
+public abstract class NonDynamicFieldMapperTestCase extends ESSingleNodeTestCase {
+
+    protected abstract String getTypeName();
+
+    protected abstract String getMapping();
```
Comment (Mikep86) [inline server/src/test/java/org/elasticsearch/index/mapper/NonDynamicFieldMapperTestCase.java:29]: Nit: The format of the string returned by this method is quite specific and could trip people up in the future. Would a `Map<String, Object>` return type here work? We could then convert it to JSON in the abstract class. WDYT?

**Review by Mikep86 — APPROVED** [review]
LGTM! Left one comment about potentially improving test ergonomics.

File: server/src/test/java/org/elasticsearch/index/mapper/NonDynamicFieldMapperTestCase.java (hunk 2 above)
Comment (carlosdelest) [reply server/src/test/java/org/elasticsearch/index/mapper/NonDynamicFieldMapperTestCase.java:29]: I've been thinking about this too - I'd like to have more use cases before reaching a decision about the format. I don't see a particular advantage on using something similar to `MapperTestCase.minimalMapping()`, at least for now.

**Pull request merged**